imblearn/metrics/pairwise.py

```diff
@@ -0,0 +1,162 @@
+"""Metrics to perform pairwise computation."""
+
+# Authors: Guillaume Lemaitre <g.lemaitre58@gmail.com>
+# License: MIT
+
+import numpy as np
+from scipy.spatial import distance_matrix
+
+
+class ValueDifferenceMetric:
+    r"""Class implementing the Value Difference Metric.
+
+    This metric computes the distance between samples containing only nominal
+    categorical features. The distance between feature values of two samples
+    is defined as:
+
+    .. math::
+       \delta(x, y) = \sum_{c=1}^{C} |p(c|x_{f}) - p(c|y_{f})|^{k} \ ,
+
+    where :math:`x` and :math:`y` are two samples and :math:`f` a given
+    feature, :math:`C` is the number of classes, :math:`p(c|x_{f})` is the
+    conditional probability that the output class is :math:`c` given that
+    the feature value :math:`f` has the value :math:`x` and :math:`k` an
+    exponent usually defined to 1 or 2.
+
+    The distance for the feature vectors :math:`X` and :math:`Y` is
+    subsequently defined as:
+
+    .. math::
+       \Delta(X, Y) = \sum_{f=1}^{F} \delta(X_{f}, Y_{f})^{r} \ ,
+
+    where :math:`F` is the number of feature and :math:`r` an exponent usually
+    defined equal to 1 or 2.
+
+    The definition of this distance was propoed in [1]_.
+
+    Parameters
+    ----------
+    classes : ndarray of shape (n_classes,)
+        The unique labels in `y`.
+
+    categories : list of arrays
+        List of arrays containing the categories for each feature. You can pass
+        the fitted attribute `categories_` of the
+        :class:`~sklearn.preprocesssing.OrdinalEncoder` used to encode the
+        data.
+
+    k : int, default=1
+        Exponent used to compute the distance between feature value.
+
+    r : int, default=2
+        Exponent used to compute the distance between the feature vector.
+
+    Attributes
+    ----------
+    proba_per_class_ : list of ndarray of shape (n_categories, n_classes)
+        List of length `n_features` containing the conditional probabilities
+        for each category given a class.
+
+    Notes
+    -----
+    The input data `X` are expected to be encoded by an
+    :class:`~sklearn.preprocessing.OrdinalEncoder` and the data type is used
+    should be `np.int32`. If other data types are given, `X` will be converted
+    to `np.int32`.
+
+    References
+    ----------
+    .. [1] Stanfill, Craig, and David Waltz. "Toward memory-based reasoning."
+       Communications of the ACM 29.12 (1986): 1213-1228.
+    """
+
+    def __init__(self, classes, categories, *, k=1, r=2):
+        self.classes = classes
+        self.categories = categories
```

Comment:
I think that we can avoid that since we require `X` and `y` in fit. The categories are actually the `X` it self, no?

Reply:
Yes but this is costly to compute and you already did the job outside using the `OrdinalEncoder`. An option might be to have `categories="auto"` and compute `np.unique` for each feature in `X` at fit?

Reply:
Yeap. This makes sense.